Case Search on Clear

Jtang-1 opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

Adds a configuration toggle under "Search and Claims Options". It is visible only when [Split Screen Case Search](https://www.commcarehq.org/hq/flags/edit/split_screen_case_search/)is enabled.
![Screen Shot 2024-01-16 at 10 32 30 AM](https://github.com/dimagi/commcare-hq/assets/88759246/acf5b6e5-acbd-4ff9-af16-90b230bc8313)

In the demo, there is a default case search `name="Jonathan"`. On clearing, it searches for cases that meets that filter.

https://github.com/dimagi/commcare-hq/assets/88759246/1e42a9c2-5260-4986-b354-32d6870da2b6



## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
[Jira Ticket](https://dimagi-dev.atlassian.net/browse/USH-4046)
[Formplayer PR](https://github.com/dimagi/formplayer/compare/jt/clear-on-search?expand=1)
[Commcare Core PR](https://github.com/dimagi/commcare-core/compare/jt/clear-on-search?expand=1)

Commits [86f807a](https://github.com/dimagi/commcare-hq/pull/33983/commits/86f807ab7b71d27f56b436524e0ee9a284bdc13c) through [e9d1b45](https://github.com/dimagi/commcare-hq/pull/33983/commits/e9d1b45c0353d8dcc3b60a9a1c3988b3a32daf85) are refactors of showing menu for split screen case search

I added validation that multi-select auto select and search on clear can't be enabled at the same time. See this [ticket](https://dimagi-dev.atlassian.net/browse/USH-3532) for incompatibility reasons. 

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
[Split Screen Case Search](https://www.commcarehq.org/hq/flags/edit/split_screen_case_search/)
## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
locally tested

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
added suite test that `search_on_clear` attribute correctly populates suite.
existing javascript test exists that checks correct views are called for split screen case search 
### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
no QA 
### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

Jtang-1 committed:
feat: adds hq ui for toggling search on clear

Jtang-1 committed:
feat: subscribe UI checkbox to save

Jtang-1 committed:
refactor: move arguements into kwargs

Jtang-1 committed:
feat: adds attribute to suite

Jtang-1 committed:
specify case where "sidebar" region is emptied

Jtang-1 committed:
mostly a refactor to pull out main region setting for split screen case search and limited to only display case list when sidebar is enabled

Previously, it the main region would show the respective menuResponse view. As long as sidebar is enabled and it is not a query view. But
this is not used for any workflows so adds unneccessary complexity

Jtang-1 committed:
refactor: pull sidebar region views into separate method

Jtang-1 committed:
refactor: moves manipulation of menuDate for caselist view to be under existing conditional

Jtang-1 committed:
feat: search on query clear when configuration is enabled

Jtang-1 committed:
bug: when response is type entity, the main region shouldn't show just the empty case list view.

The change also avoids populating "main" region twice

Jtang-1 committed:
bug: don't search on clear when screen is small since query view is a popup and not perpetually displayed at the same time as entity view

Jtang-1 committed:
test: tests search_on_clear is populated in suite

Jtang-1 committed:
adds validation for incompatibility of autoselect with search on clear

when both are configured, it will cause an infinite rerunning of searches in a loop

Jtang-1 committed:
feat: hides config if dynamic search is enabled

dynamic search includes "search on clear" behavior

stephherbers commented on corehq/apps/app_manager/templates/app_manager/partials/modules/case_search_properties.html:
```diff
@@ -466,11 +466,23 @@ <h4 class="panel-title panel-title-nolink">{% trans "Search and Claim Options" %
             </div>
           </div>
           {% endif %}
+          {% if request|toggle_enabled:'SPLIT_SCREEN_CASE_SEARCH' and not app.split_screen_dynamic_search%}
```
why would this not be configurable with dynamic search?

Jtang-1 replied:
Dynamic search already includes "search on clear" behavior. So if dynamic search is enabled, there's no need to also specifically enabled "search on clear" so the UI toggle is hidden.

stephherbers replied:
cool I missed that detail sounds good

stephherbers commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:
```diff
@@ -584,6 +584,7 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
 
             this.dynamicSearchEnabled = !(options.disableDynamicSearch || this.smallScreenEnabled) &&
                 (toggles.toggleEnabled('DYNAMICALLY_UPDATE_SEARCH_RESULTS') && this.options.sidebarEnabled);
+            this.searchOnClear = (options.searchOnClear && !this.smallScreenEnabled);
```
why is this not enabled with small screens?

Jtang-1 replied:
Correct me if i'm wrong, but the reason dynamic search isn't enabled on small screen is because the query view and case list view aren't visible at the same time. This is the same rationale as to why I also disabled "search on clear" for small screens

stephherbers replied:
cool yep that's right sounds good

stephherbers commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/controller.js:
```diff
@@ -106,26 +106,44 @@ hqDefine("cloudcare/js/formplayer/menus/controller", function () {
+        }
+    };
+
+    var showSplitScreenQuery = function (menuResponse, menuListView) {
```
⭐️ great refactor

stephherbers commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/controller.js:
```diff
@@ -106,26 +106,44 @@ hqDefine("cloudcare/js/formplayer/menus/controller", function () {
-            var caseListView = menusUtils.getCaseListView(menuResponse);
-            FormplayerFrontend.regions.getRegion('main').show(caseListView(menuData));
-        } else if (menuListView) {
+        if (menuListView && !sidebarEnabled) {
```
can this be simplified to just `!sidebarEnabled`

Jtang-1 replied:
the very initial implementation checks `menuListView` isn't null so I just kept it. I also think it makes sense since `getMenuView` can potentially return null.

stephherbers commented on corehq/apps/app_manager/templates/app_manager/partials/build_errors.html:
```diff
@@ -195,6 +195,11 @@ <h4 class="alert-heading">
               {% blocktrans with module_name=error.module.name|trans:langs %}
               The case list in <a href="{{ module_url }}">{{ module_name }}</a> can not use the same "search input instance name" as its Parent Select Menu.
               {% endblocktrans %}
+            {% case "search on clear with auto select" %}
```
add test for this case

Jtang-1 replied:
[b52eced](https://github.com/dimagi/commcare-hq/pull/33983/commits/b52eced512c8c1dcbdcb918df790aa6e5e0f607d)

stephherbers reviewed (commented):
great foundation just questions on compatibility

Jtang-1 committed:
test: validation error is triggered when app configured with both "search on clear" and auto select

stephherbers commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:
```diff
@@ -705,7 +706,7 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
             });
             self.setStickyQueryInputs();
-            if (self.dynamicSearchEnabled) {
+            if (self.dynamicSearchEnabled || this.searchOnClear) {
```
nit: if `searchOnClear` is only true when `dynamicSearchEnabled` is enabled, then this can just be `dynamicSearchEnabled`

stephherbers replied:
gotcha-- cool makes sense. Ignore this comment then

stephherbers commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js:
```diff
@@ -584,6 +584,7 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
 
             this.dynamicSearchEnabled = !(options.disableDynamicSearch || this.smallScreenEnabled) &&
                 (toggles.toggleEnabled('DYNAMICALLY_UPDATE_SEARCH_RESULTS') && this.options.sidebarEnabled);
+            this.searchOnClear = (options.searchOnClear && !this.smallScreenEnabled);
```
should this check for `this.searchOnClear = (options.searchOnClear && this.dynamicSearchEnabled);` to clarify requirement and add an extra check?

Jtang-1 replied:
no, same reason as this [comment](https://github.com/dimagi/commcare-hq/pull/33983#discussion_r1456332158)

But it is dependent on Split screen case search FF. I haven't quite figured out where the FF checks should happen as opposed to just putting it in every step. Right now, with these steps
1. SSCS enabled and searchOnClear enabled -> FP menu response sets "searchOnClear" as True
2. SSCS is disabled, search_on_clear is still defined as True on module model -> FP menu response still sets "searchOnClear" as True

I'm thinking I should add the FF check on FP side, otherwise menuResponse is wrong. Do you think that makes sense?

Jtang-1 replied:
I decided to do a check before populating the suite [2168b86](https://github.com/dimagi/commcare-hq/pull/33983/commits/2168b86fc9f306bf50a6e00792f8d0797338c0d9). It'll require app rebuild but I think that makes sense as that's how most of our FF operates.

stephherbers replied:
agreed on approach of checking before populating the suite to keep the logic within this repo.

stephherbers reviewed (approved):
added non-blocking comments

Jtang-1 committed:
bug: search_on_clear should only be enabled Split screen case search is also enabled

If module has search_on_clear enabled, but then turns of SSCS FF,
previously, formplayer menuResponse would still respond searchOnClear = True. With this change,
response is false for searchOnClear

stephherbers commented on corehq/apps/app_manager/templates/app_manager/partials/modules/case_search_properties.html:
```diff
@@ -466,11 +466,23 @@ <h4 class="panel-title panel-title-nolink">{% trans "Search and Claim Options" %
             </div>
           </div>
           {% endif %}
+          {% if request|toggle_enabled:'SPLIT_SCREEN_CASE_SEARCH' and not app.split_screen_dynamic_search%}
```
should a check be added to the dynamic search checkbox to check that `app.search_on_clear` is not enabled?

Jtang-1 replied:
I don't think so because dynamic search has behaviors in addition to searching on clear. So app builder may decide to enable dynamic search even after enabling "search on clear" which would be fine.

orangejenny committed:
Merge branch 'master' into jt/case_search_on_clear

orangejenny committed:
Fixed bad merge

Jtang-1 committed:
Merge branch 'master' into jt/case_search_on_clear

stephherbers commented:
offlined with Jonathan and if things looked good I would merge since he's offline today.
